File: src/day1/Stack.ts
type Node<T> = {
    value: T;
    next?: Node<T> | undefined;
};

export default class Stack<T> {
    public length: number;
    private _tail: Node<T> | undefined;

    constructor() {
        this.length = 0;
        this._tail = undefined;
    }

    push(item: T): void {
        const node: Node<T> = {
            value: item,
        };
        this.length++;
        if (!this._tail) {
            this._tail = node;
        } else {
            node.next = this._tail;
            this._tail = node;
        }
    }

    pop(): T | undefined {
        if (!this._tail) {
            return;
        }
        const oldTail = this._tail;
        this._tail = this._tail.next;
        this.length--;
        return oldTail.value;
    }

    peek(): T | undefined {
        return this._tail?.value;
    }
}

File: src/golang/stack.go
package main

type Node[T any] struct {
	data T
	next *Node[T]
}

type Stack[T any] struct {
	tail   *Node[T]
	Length int
}

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{}
}

func (s *Stack[T]) Peek() T {
	return s.tail.data
}

func (s *Stack[T]) Push(item T) {
	node := &Node[T]{
		data: item,
	}
	s.Length++
	node.next = s.tail
	s.tail = node
}

func (s *Stack[T]) Pop() T {
	var zero T
	if s.tail == nil {
		return zero
	} else {
		s.Length--
		oldHead := s.tail
		s.tail = s.tail.next
		return oldHead.data
	}
}

func main() {
}
